### PR TITLE
0.18.1: Backports

### PR DESCRIPTION
**Pull request by maflcko**

### CODE REVIEW

**Comment by laanwj** [comment]
utACK for 5935f0126ef37175a8fbfee017b470af13aad46d

But I guess it makes sense to keep this open for a while for other things that need to be backported to the 0.18 branch.

**Comment by fanquake** [comment]
Could include #14984.

**Comment by promag** [comment]
@MarcoFalke I won't open a PR to backport 14984 until you say so, it should be a clean cherry pick.

**Comment by promag** [comment]
Can you include #16122 and #16135?

**Comment by MarcoFalke** [comment]
Done

**Comment by fanquake** [comment]
The GCC bug test backport needs to be fixed for the `setup_common` refactor.
```shell
test/compilerbug_tests.cpp:5:10: fatal error: test/setup_common.h: No such file or directory
 #include <test/setup_common.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by MarcoFalke** [comment]
Ugh, thx. Fixed.

**Comment by fanquake** [comment]
Can add #15899.

**Comment by MarcoFalke** [comment]
This is ready for review/merge

**Comment by fanquake** [comment]
Can you add #16231 as well? I will review after that.

**Comment by practicalswift** [comment]
@MarcoFalke Why is makes #16205 important to backport?

**Comment by fanquake** [comment]
@practicalswift It's mentioned at the bottom of that PR: https://github.com/bitcoin/bitcoin/pull/16205#issuecomment-503131367.

> Looks like this accidentally fixed multiple test failures with the cross-compiled windows binaries
> I am going to backport this, since it turned out to be a bugfix

**Comment by MarcoFalke** [comment]
Thanks for asking. Please let me know if there is a pull request missing for backport or something was backported in error.

Apart from a bunch one-line conflicts, the commits are all clean cherry-picks.

File: test/functional/p2p_segwit.py
Comment (fanquake) [inline test/functional/p2p_segwit.py]:
```bash
Traceback (most recent call last):
  File "/Users/michael/github/bitcoin/test/functional/test_framework/test_framework.py", line 175, in main
    self.run_test()
  File "/Users/michael/github/bitcoin/test/functional/p2p_segwit.py", line 278, in run_test
    self.test_superfluous_witness()
  File "/Users/michael/github/bitcoin/test/functional/p2p_segwit.py", line 2076, in test_superfluous_witness
    assert_raises_rpc_error(-22, "TX decode failed", self.nodes[0].decoderawtransaction, serialize_with_bogus_witness(tx).hex())
AttributeError: 'bytes' object has no attribute 'hex'
```

**Review by fanquake — COMMENTED** [review]
Checked 5935f0126ef37175a8fbfee017b470af13aad46d and ensured that `-fstack-reuse=none` was being added to CXXFLAGS when compiling with GCC:
```bash
  CC            = gcc
  CFLAGS        = -g -O2
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = g++ -std=c++11
  CXXFLAGS      =   -fstack-reuse=none -Wstack-protector -fstack-protector-all  -Wall -Wextra -Wformat -Wvla -Wredundant-decls  -Wno-unused-parameter -Wno-implicit-fallthrough   -g -O2
  LDFLAGS       = -pthread  -Wl,-z,relro -Wl,-z,now -pie  
  ARFLAGS       = cr
...
(16035)root@69972490e9d9:/bitcoin# gcc --version
gcc (Debian 8.3.0-6) 8.3.0
```

Checked 3dbc7def0f12257918b87e6abd6432d40d3e59f9, and that loaded wallets are shown as disabled.

loaded_wallets

Checked b55cbe82d98338c3c63770d624bda64cb0f472b9 and that the `bitcoin.conf` opening behaviour works as expected.

bitcoin_conf

Checked 7ed1a601934625a8073324da6e3ec881cec01e05 and that it’s not possible to enter and execute commands in the console during startup.

Checked d80c558e026a4e500d5894d0c386378ac7a4d20f and that you cannot cause a segfault using:
```bash
src/bitcoin-cli -regtest dumpwallet ~/downloads/wallet
src/bitcoin-cli -regtest importwallet ~/downloads/wallet
```

Checked d1f261150b49a957138ef4920203f91f5a8c01b6. Note that this is not a clean cherry-pick. `#include <test/setup_common.h>` has been fixed up to be `#include <test/test_bitcoin.h>` for `0.18`.

Also checked that the test failed if you compile without `-fstack-protector=none`:
```bash
Running 1 test case...
Entering test module "Bitcoin Test Suite"
test/compilerbug_tests.cpp(8): Entering test suite "compilerbug_tests"
test/compilerbug_tests.cpp(30): Entering test case "gccbug_90348"
test/compilerbug_tests.cpp(38): error: in "compilerbug_tests/gccbug_90348": check check_zero(in, i) has failed
test/compilerbug_tests.cpp(38): error: in "compilerbug_tests/gccbug_90348": check check_zero(in, i) has failed
test/compilerbug_tests.cpp(38): error: in "compilerbug_tests/gccbug_90348": check check_zero(in, i) has failed
test/compilerbug_tests.cpp(38): error: in "compilerbug_tests/gccbug_90348": check check_zero(in, i) has failed
test/compilerbug_tests.cpp(30): Leaving test case "gccbug_90348"; testing time: 19263us
test/compilerbug_tests.cpp(8): Leaving test suite "compilerbug_tests"; testing time: 19286us
Leaving test module "Bitcoin Test Suite"; testing time: 19364us
*** 4 failures are detected in the test module "Bitcoin Test Suite"
make[3]: *** [Makefile:14472: test/compilerbug_tests.cpp.test] Error 1
```

Note that 592016ba18f2fbcb382ac0306ae89aba1b4be843 is not a clean cherry-pick. It has the removal of the `assert_greater_than` include, which is not in the referenced commit (f402012ccfc596d7d94851dabbf386c278ff5335).

Note that 79745d1752d0d1a95f073a04d6f4b9715745cf17 does not contain the src/sync.cpp change from fa8f195195945ce6258199af0461e3fbfbc1236d.

I assume d9fc969e71c1f0f0a2404c3bb08aad78b6ac7a39 was backported because other test changes rely on it? Regardless, why not backport the other commit (b67978529ad02fc2665f2362418dc53db2e25e17) from # 15826? I’d have thought backporting the documentation would also be useful.

Checked 2800b3d5c14a9e5437b7e8838c69fdd6710908bb and that wallets still appear in the GUI.

listed_wallets

Checked 715da91e917af106d8d77fb6240f2adf84bca292 and that the warning is removed on a Ubuntu Bionic system. Given I'm running in virtualbox, I don't think I can see the same improvements that others saw in #16254.

ubuntu

Other commits I looked over the backported code.

Currently seeing failures running the functional tests. Have commented inline.

File: test/functional/p2p_segwit.py
Comment (MarcoFalke) [reply test/functional/p2p_segwit.py]: @fanquake `.hex` requires python3.5

Comment (fanquake) [reply test/functional/p2p_segwit.py]: @MarcoFalke isn’t the 0.18 branch minimum 3.4.9?

**Comment by MarcoFalke** [comment]
>  assume d9fc969 was backported because other test changes rely on it? Regardless, why not backport the other commit (b679785) from # 15826? I’d have thought backporting the documentation would also be useful.

Indeed. Good point on the doc, added as backport.

File: test/functional/p2p_segwit.py
Comment (MarcoFalke) [reply test/functional/p2p_segwit.py]: Bumped it, since 3.4 is EOL

**Review by promag — COMMENTED** [review]
Commit 79745d1752d0d1a95f073a04d6f4b9715745cf17 is missing:
```diff
--- a/src/sync.cpp
+++ b/src/sync.cpp
@@ -47,7 +47,9 @@ struct CLockLocation {

     std::string ToString() const
     {
-        return mutexName + "  " + sourceFile + ":" + itostr(sourceLine) + (fTry ? " (TRY)" : "");
+        return strprintf(
+            "%s %s:%s%s (in thread %s)",
+            mutexName, sourceFile, itostr(sourceLine), (fTry ? " (TRY)" : ""), m_thread_name);
     }

 private:
```
Is there a reason?

Edit: if the goal is to just replace `fprintf` then it's all good.

**Comment by promag** [comment]
ACK bcb27d7, I've verified all commits but just built each unclean cherry pick (which are trivial).

**Review by fanquake — APPROVED** [review]
ACK bcb27d7b03d1e7a4404c04859f8146d8bc7c360a

Had a quick look over af25a757e077f907a8ce1cb676bd4908ed638e14, and confirmed that the functional tests are now running (using 3.5.6) after bcb27d7b03d1e7a4404c04859f8146d8bc7c360a. Full test output is [here](https://gist.github.com/fanquake/20fa38206251a4807eb154596f11df3e). Note that `p2p_invalid_messages.py` fails quite often on macOS, so isn't an issue here.

I agree that this is turning into a lot to review and test, so would suggest we merge this soon (assuming some more reviewers) without piling to much more in.

**Comment by luke-jr** [comment]
These should be a clean merge, not cherry-picked:

```
Bugfix: test/functional/rpc_psbt: Remove check for specific error mes…
Bugfix: test/functional/rpc_psbt: Correct test description comment ￼…
```

**Review by luke-jr — CHANGES_REQUESTED** [review]
These should be a clean merge, not cherry-picked:

```
Bugfix: test/functional/rpc_psbt: Remove check for specific error mes…
Bugfix: test/functional/rpc_psbt: Correct test description comment ￼…
```

**Comment by MarcoFalke** [comment]
@luke-jr I used cherry-picks for consistency. (Most other backports are cherry-picks)

**Comment by luke-jr** [comment]
Merges should be used whenever possible. (Ideally, it would be nice if we could get to the point where most fixes are clean merges.)

**Comment by MarcoFalke** [comment]
I agree, but sometimes it is not clear in advance what is a fix and needs to be backported. Sounds off-topic for this pull request and could be raised in an irc meeting?

**Comment by luke-jr** [comment]
The incorrect cherry-picks are part of this PR :)

**Comment by MarcoFalke** [comment]
This has been the policy for years, afaic. So to change it, it would have to be discussed. Also, I am not sure how that will interfere with the release process scripts of @laanwj.

**Comment by luke-jr** [comment]
It was never a policy. For a few recent examples, #15552 and #15524 were merged into 0.18 cleanly.

**Review by SkippydooDougstrong7772 — APPROVED** [review]

**Comment by Nomed456** [comment]
P

On 1 Jul 2019 19:07, "SkippydooDougstrong7772" <notifications@github.com>
wrote:

*@SkippydooDougstrong7772* approved this pull request.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bitcoin/bitcoin/pull/16035?email_source=notifications&email_token=AAS3KFIWTWSLRMC4B4PHVCLP5I2WJA5CNFSM4HNM6JPKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5EUKQQ#pullrequestreview-256460098>,
or mute the thread

**Comment by ajtowns** [comment]
ACK bcb27d7b03d1e7a4404c04859f8146d8bc7c360a ; compiled and ran the tests; checked the commits match the commits they claim to, and that those commits are in master. didn't check that the commits haven't been reverted in master.

**Comment by laanwj** [comment]
ACK bcb27d7b03d1e7a4404c04859f8146d8bc7c360a

Please don't merge yet as @fanquake would like to do this as his first merge to a branch :smile: 

> This has been the policy for years, afaic. So to change it, it would have to be discussed. Also, I am not sure how that will interfere with the release process scripts of @laanwj.

It can handle either merges (only requirement is to have the correct PR # in the title of the merge commit), or the `'Github-Pull: #?(\d+)'` metadata.

**Pull request merged**
